From af40a33f960c264ba41907c806c56864de06c451 Mon Sep 17 00:00:00 2001
From: Li Nan <linan122@huawei.com>
Date: Tue, 16 Aug 2022 14:57:42 +0800
Subject: [PATCH] cgroup: Support iocost for cgroup v1
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: af40a33f960c264ba41907c806c56864de06c451
Modified-by-SEL: No


hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I63YN6

---------------------------

Since iocost can only be used in cgroup v2, add it to v1 now.

Signed-off-by: Li Nan <linan122@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 block/blk-iocost.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/block/blk-iocost.c b/block/blk-iocost.c
index 08e4ba856e3b..64119eaf29da 100644
--- a/block/blk-iocost.c
+++ b/block/blk-iocost.c
@@ -3396,6 +3396,28 @@ static ssize_t ioc_cost_model_write(struct kernfs_open_file *of, char *input,
 	return ret;
 }
 
+static struct cftype ioc_legacy_files[] = {
+	{
+		.name = "cost.weight",
+		.flags = CFTYPE_NOT_ON_ROOT,
+		.seq_show = ioc_weight_show,
+		.write = ioc_weight_write,
+	},
+	{
+		.name = "cost.qos",
+		.flags = CFTYPE_ONLY_ON_ROOT,
+		.seq_show = ioc_qos_show,
+		.write = ioc_qos_write,
+	},
+	{
+		.name = "cost.model",
+		.flags = CFTYPE_ONLY_ON_ROOT,
+		.seq_show = ioc_cost_model_show,
+		.write = ioc_cost_model_write,
+	},
+	{}
+};
+
 static struct cftype ioc_files[] = {
 	{
 		.name = "weight",
@@ -3420,6 +3442,7 @@ static struct cftype ioc_files[] = {
 
 static struct blkcg_policy blkcg_policy_iocost = {
 	.dfl_cftypes	= ioc_files,
+	.legacy_cftypes = ioc_legacy_files,
 	.cpd_alloc_fn	= ioc_cpd_alloc,
 	.cpd_free_fn	= ioc_cpd_free,
 	.pd_alloc_fn	= ioc_pd_alloc,
-- 
2.33.0

